text/0000-guard_clause_flow_typing.md

```diff
@@ -0,0 +1,284 @@
+  if let Ok(f) = file {
+    let mut buf_reader = BufReader::new(f);
+    let mut contents = String::new();
+
+    if buf_reader.read_to_string(&mut contents).is_ok() {
+      let mut data: Vec<u8> = vec![];
+
+      for item in contents.
+          split("\n").
+          map(|s| s.to_string()).
+          filter(|s| !s.is_empty()).
+          collect::<Vec<String>>() {
+
+        let num = item.parse::<u8>();
+
+        if let Ok(conf) = num {
+          data.push(conf);
+        } else {
+          return Err(Error::ConfigParseFail);
+        }
+      }
+
+      Ok( Config { data: data } )
+    } else {
+      Err(Error::ConfigLoadFail)
+    }
+  } else {
+    Err(Error::ConfigLoadFail)
+  }
+}
+```
+### Wrong (B)
+```rust
+fn read_config2() -> Result<Config, Error> {
+  let file = File::open("program.cfg");
+  match file {
+    Ok(f) => {
+      let mut buf_reader = BufReader::new(f);
+      let mut contents = String::new();
+
+      match buf_reader.read_to_string(&mut contents) {
+        Ok(_) => {
+          let mut data: Vec<u8> = vec![];
+
+          for item in contents.
+              split("\n").
+              map(|s| s.to_string()).
+              filter(|s| !s.is_empty()).
+              collect::<Vec<String>>() {
+
+            let num = item.parse::<u8>();
+
+            match num {
+              Ok(conf) => data.push(conf),
+              _ => { return Err(Error::ConfigParseFail); },
+            }
+          }
+
+          Ok( Config { data: data } )
+        },
+        _ => { Err(Error::ConfigLoadFail) }
+      }
+    },
+    _ => { Err(Error::ConfigLoadFail) }
+  }
+}
+```
+### Wrong (C)
+```rust
+fn read_config3() -> Result<Config, Error> {
+  let file = File::open("program.cfg");
+
+  if let Ok(f) = file {
+    let mut buf_reader = BufReader::new(f);
+    let mut contents = String::new();
+
+    if buf_reader.read_to_string(&mut contents).is_ok() {
+      let mut data: Vec<u8> = vec![];
+
+      for item in contents.
+          split("\n").
+          map(|s| s.to_string()).
+          filter(|s| !s.is_empty()).
+          collect::<Vec<String>>() {
+
+        let num = item.parse::<u8>();
+
+        if let Ok(conf) = num {
+          data.push(conf);
+        } else {
+          return Err(Error::ConfigParseFail);
+        }
+      }
+
+      return Ok( Config { data: data } );
+    }
+  }
+
+  Err(Error::ConfigLoadFail)
+}
+```
+
+And here is the correct usage of a _Guard Clause_ which allows us to avoid deeply nested logic.
+
+### Correct
+```rust
+fn read_config4() -> Result<Config, Error> {
+  let file = File::open("program.cfg");
+
+  // Correct use of Guard Clause
+  if let Err(_) = file { return Err(Error::ConfigLoadFail); }
+  
+  // Safe use of pattern matching assignment after Guard Clause
+  let Ok(f) = file;
+
+  let mut buf_reader = BufReader::new(f);
+  let mut contents = String::new();
+
+  // Correct use of Guard Clause
+  if let Err(_) = buf_reader.read_to_string(&mut contents) {
```

Comment:
This is already valid Rust.

Reply:
Correct.

Reply:
@sfackler I think the key is that the following line would currently be an "irrefutable pattern" error:

https://github.com/rust-lang/rfcs/pull/2221/files#diff-5935e1ae73da8f77fd623e86bd4eb1eeR159

Reply:
I get the point of this is supposed to provide alternatives to map_err and stuff, but I find that this:

```rust
let file = File::open("program.cfg").map_err(|_| Error::ConfigLoadFail)?;
```
is far more readable and a one liner than:

```rust
let file = File::open("program.cfg");
if let Err(_) = file { return Err(Error::ConfigLoadFail); }
let Ok(f) = file;
```

It kind of obscures what's going on and you mentioned this was going to help new users, but I find it more confusing and less clear as to what's going on. This guard clause feels like a weird version of ?/try!(). I'd expect new users to use copious amounts of `match`, ask why they have so many nested clauses, then get directed to ?/try!() and all of Result's methods for dealing with these, which has happened before and most new users go "Oh okay thanks" and go on about their day and start using things like `map_err`.

Most of the above examples above given as an example can easily be rewritten to avoid nested if/else clauses.
For example Wrong A can be rewritten:

```rust
fn read_config1() -> Result<Config, Error> {
    let file = File::open("program.cfg").map_err(|_| Error::ConfigLoadFail)?;
    let mut buf_reader = BufReader::new(f);
    let mut contents = String::new();
    buf_reader.read_to_string(&mut contents).map_err(|_| Error::ConfigLoadFail)?
    let mut data: Vec<u8> = Vec::new();
    for item in contents
                      .split("\n")
                      .map(|s| s.to_string())
                      .filter(|s| !s.is_empty())
                      .collect::<Vec<String>>()
    {
        Ok(Config { 
             data: data.push(item.parse::<u8>().map_err(|_| Error::ConfigParseFail)?) 
         } )
    }
}
```

Which makes the function shorter but is far more readable as well. If the goal is to teach new people Rust, then we should be focusing on the documentation to cover this and make it less confusing if there is some, because this is considered more idiomatic. I don't think we need more ways to do match.

Reply:
@mark-i-m That's not the line following this comment.

Reply:
@sfackler I meant the line with that I linked... It seemed like it would be confusing to add a new comment on that line instead... sorry for the confusion.

Reply:
@mgattozzi In your revised code sample isn't the for loop generating a new `Result<Config>` through each iteration?  And the ownership of `data` would only work if the underlying type is `Copy` here right?  I've never seen **for** used in this way (generating & ignoring extra instances).